src/dbnode/storage/database.go

```diff
@@ -542,11 +565,20 @@ func (d *db) queueBootstrapWithLock() {
 	// the non-clustered database bootstrapped by assigning it shardsets which will trigger new
 	// bootstraps since d.bootstraps > 0 will be true.
 	if d.bootstraps > 0 {
+		d.log.Info("enqueuing bootstrap")
 		bootstrapAsyncResult := d.mediator.BootstrapEnqueue()
-		// NB(linasn): We need to wait for the bootstrap to start and set it's state to Bootstrapping in order
-		// to safely run fileOps in mediator later.
-		bootstrapAsyncResult.WaitForStart()
+		go func() {
+			bootstrapAsyncResult.WaitForComplete()
+			// NB(linasn): We don't want to invoke onCompleteFn if another bootstrap is already
+			// started because we are actually not initiating new bootstrap.
+			if !bootstrapAsyncResult.bootstrapResult.AlreadyBootstrapping {
+				onCompleteFn()
```

Comment:
So I'm a little concerned that `onCompleteFn()` doesn't get called sometimes and then if the `enqueueBootstrapWithLock` is not called with a completion function then we will never re-enable file ops.

I think we should make `BootstrapEnqueue` take the completion function so that it's guaranteed to execute once all bootstraps have completed.

Reply:
Merged changes from your branch so the completion function is now passed to the bootstrap method itself.